presto-main/src/main/java/com/facebook/presto/cost/CostCalculatorWithEstimatedExchanges.java

```diff
@@ -0,0 +1,194 @@
+            this.stats = requireNonNull(stats, "stats is null");
+            this.lookup = requireNonNull(lookup, "lookup is null");
+        }
+
+        @Override
+        protected PlanNodeCostEstimate visitPlan(PlanNode node, Void context)
+        {
+            // TODO implement logic for other node types and return UNKNOWN_COST here (or throw)
+            return ZERO_COST;
+        }
+
+        @Override
+        public PlanNodeCostEstimate visitGroupReference(GroupReference node, Void context)
+        {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public PlanNodeCostEstimate visitAggregation(AggregationNode node, Void context)
+        {
+            PlanNodeCostEstimate remoteRepartitionCost = CostCalculatorUsingExchanges.calculateExchangeCost(
+                    numberOfNodes,
+                    getStats(node.getSource()),
+                    REPARTITION,
+                    REMOTE);
+            PlanNodeCostEstimate localRepartitionCost = CostCalculatorUsingExchanges.calculateExchangeCost(
+                    numberOfNodes,
+                    getStats(node.getSource()),
+                    REPARTITION,
+                    LOCAL);
+            return remoteRepartitionCost.add(localRepartitionCost);
+        }
+
+        @Override
+        public PlanNodeCostEstimate visitJoin(JoinNode node, Void context)
+        {
+            return calculateJoinCost(
+                    node.getLeft(),
+                    node.getRight(),
+                    node.getDistributionType().orElse(JoinNode.DistributionType.PARTITIONED).equals(JoinNode.DistributionType.REPLICATED));
+        }
+
+        @Override
+        public PlanNodeCostEstimate visitSemiJoin(SemiJoinNode node, Void context)
+        {
+            return calculateJoinCost(
+                    node.getSource(),
+                    node.getFilteringSource(),
+                    node.getDistributionType().orElse(SemiJoinNode.DistributionType.PARTITIONED).equals(SemiJoinNode.DistributionType.REPLICATED));
+        }
+
+        private PlanNodeCostEstimate calculateJoinCost(PlanNode probe, PlanNode build, boolean replicated)
+        {
+            if (replicated) {
+                PlanNodeCostEstimate replicateCost = CostCalculatorUsingExchanges.calculateExchangeCost(
+                        numberOfNodes,
+                        getStats(build),
+                        REPLICATE,
+                        REMOTE);
+                // cost of the copies repartitioning is added in CostCalculatorUsingExchanges#calculateJoinCost
+                PlanNodeCostEstimate localRepartitionCost = CostCalculatorUsingExchanges.calculateExchangeCost(
+                        numberOfNodes,
+                        getStats(build),
+                        REPARTITION,
+                        LOCAL);
+                return replicateCost.add(localRepartitionCost);
+            }
+            else {
```

Comment:
redundant `else` block

Reply:
:( it looks so better _with_ this `else`, since the branches are _so_ equivalent..